File: java/src/P013/RomanToInteger/RomanToInteger.java
package P013.RomanToInteger;

/*
Given a roman numeral, convert it to an integer.

Input is guaranteed to be within the range from 1 to 3999.
 */

import java.util.HashMap;
import java.util.Map;

public class RomanToInteger {
    public static int romanToInt(String s) {
        if (s == null || s.length() == 0) return 0;

        Map<Character, Integer> roman = new HashMap<Character, Integer>() {{
           put('I', 1);
           put('V', 5);
           put('X', 10);
           put('L', 50);
           put('C', 100);
           put('D', 500);
           put('M', 1000);
        }};


        int num = roman.get(s.charAt(s.length()-1));
        for (int i = s.length()-2; i >= 0; i--) {
            int curr = roman.get(s.charAt(i));
            int right = roman.get(s.charAt(i+1));
            if (curr < right) num -= curr;
            else num += curr;
        }

        return num;
    }

    public static void main(String[] args) {
        System.out.println(romanToInt("MMMCMXCIX"));
    }
}

File: python/P013_Roman_to_Integer/RomanToInteger.py
# coding=utf-8


class Solution(object):
    @staticmethod
    def romanToInt(s):
        """
        Given a roman numeral, convert it to an integer.

        Input is guaranteed to be within the range from 1 to 3999.
        """
        if s is None or len(s) == 0:
            return 0

        roman = {
            'I': 1,
            'V': 5,
            'X': 10,
            'L': 50,
            'C': 100,
            'D': 500,
            'M': 1000,
        }

        s = list(map(lambda x: roman[x], s))
        pairs = list(zip(s[:-1], s[1:]))
        num = 0
        for curr, right in pairs:
            if curr >= right:
                num += curr
            else:
                num -= curr

        return num + s[-1]


if __name__ == '__main__':
    print(Solution.romanToInt('MMMCMXCIX'))